### PR TITLE
Added a custom allocator.

### PR DESCRIPTION
**Pull request by AWoloszyn**
This allows protecting all memory that was allocated using the
allocator.

### CODE REVIEW

File: core/memory/arena/cc/arena_test.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+TEST_P(memory_protection_tests, unprotect_memory) {
+  Arena a;
+  uint8_t* x = static_cast<uint8_t*>(a.allocate(GetParam(), 1));
+  *x = 4;
+  a.protect();
+  a.unprotect();
+  *x = 5;
+}
+
+// DO NOT add another death test after 8192
```
Comment (ben-clayton) [inline core/memory/arena/cc/arena_test.cpp]: Remove `after 8192`?

File: core/memory/arena/cc/arena.cpp [hunk 2]
```diff
@@ -19,91 +19,321 @@
...
+  ::free(val);
+#endif
+}
+
+void protect_range(void* val, uint32_t size) {
+#if TARGET_OS == GAPID_OS_WINDOWS
+  DWORD old_protect = 0;
+  VirtualProtect(val, size, PAGE_READONLY, &old_protect);
+#else
+    mprotect(val, size, PROT_READ);
```
Comment (ben-clayton) [inline core/memory/arena/cc/arena.cpp]: nit: tabbing

File: core/memory/arena/cc/arena.cpp [hunk 3]
```diff
@@ -19,91 +19,321 @@
...
+#endif
+}
+
+void unprotect_range(void* val, uint32_t size) {
+#if TARGET_OS == GAPID_OS_WINDOWS
+  DWORD old_protect = 0;
+  VirtualProtect(val, size, PAGE_READWRITE, &old_protect);
+#else
+    mprotect(val, size, PROT_READ | PROT_WRITE);
```
Comment (ben-clayton) [inline core/memory/arena/cc/arena.cpp]: nit: tabbing

Comment (AWoloszyn) [reply core/memory/arena/cc/arena.cpp]: Done

File: core/memory/arena/cc/arena.cpp (hunk 2 above)
Comment (AWoloszyn) [reply core/memory/arena/cc/arena.cpp]: Done

File: core/memory/arena/cc/arena_test.cpp (hunk 1 above)
Comment (AWoloszyn) [reply core/memory/arena/cc/arena_test.cpp]: Done

**Pull request merged**